Dynamic Auth proposal

guicassolato opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Extends the `AccessPolicy` API for more dynamic agentic scenarios such as: unbounded number of identities, identity federation, authorization based on common patterns of the targeted resources and/or authorization decision offloaded to external systems.

The new fields added to the API can used in conjunction or alternatively to the existing ones which continue to allow defining static, inline access control rules.

**Does this PR introduce a user-facing change?**:

```release-note
Extends the `AccessPolicy` API for more dynamic agentic scenarios such as: unbounded number of identities, identity federation, authorization based on common patterns of the targeted resources and/or authorization decision offloaded to external systems.
```

david-martin commented:
@guicassolato I've been trying to get my head around what vendors would implement vs. how they might extend this with additional capabilities.
I put together a simple relationship diagram so I could better understand how things are proposed right now:

<img width="2407" height="1661" alt="image" src="https://github.com/user-attachments/assets/b0cafab6-1dcd-4d96-934e-82eb9e7b51b4" />

I believe the proposal would require implementors to implement for Kube & OIDC for identity, and Kube & CEL for authorization. That would align with some future conformance potentially.
If an implementer wants to add additional identity rule or authorization rule types, they can, but they wouldn't be part of the standard. At least not without updating the corresponding type (IdentityRuleType/AuthorizationRuleType) and doing a new release of the API.


I've drawn up another version of this diagram, with an identity provider and authorization policy reference from the corresponding fields in the AuthRule. My intent is to abstract the points of contention to a clearer vendor extension point.
In this approach, an IdentityRule no longer has a type. Instead it has a `providerRef`. Same for AuthorizationRule. This is the extension point where a vendor can implement their own supported providers, like OIDC, Kube, CEL etc.. We could still build these out in a reference implementation, but they wouldn't be part of the standard API.

<img width="2173" height="1840" alt="image" src="https://github.com/user-attachments/assets/df0fc8b9-6444-4b44-a1b4-fc640817a387" />

For simpler cases, there is a provision in the spec for an inline rule. This could be more like the AccessRule in https://github.com/kubernetes-sigs/kube-agentic-networking/blob/main/docs/proposals/0008-ToolAuthAPI.md, where a simple list of tools or identities is provided inline.

guicassolato commented on docs/proposals/0017-DynamicAuth.md:
```diff
@@ -0,0 +1,537 @@
+}
+
+// IdentityRuleType defines a type of identity verification rule
+// +kubebuilder:validation:Enum=Kubernetes,OIDC
```
@howardjohn, I see your point and I don't disagree with it. However, I also believe this is part of building a standard.

The standard can evolve to supporting more types in the future. We just have to start somewhere.

I guess the issue then is which types should we start with, that are better than having no standard types at all? Or, as you said, why Kube auth and OIDC, and not LDAP, SAML, Kerberos, Tailscale, etc?

My case for these two (namely, Kube auth and OIDC) _as a starting point_ is backed by experience, but that matters less than the following reasons, I suppose:
- Kubernetes is the platform, it is the only given we have. Using the authz system provided by the platform makes sense to me. You don't need a 3rd-party authz system, you know the language already, etc, etc.
- OIDC not only is widely popular but it is the _de facto_ authentication system in MCP, which, as we know, standardised on OAuth 2.1.

guicassolato replied:
@howardjohn, do you think [this](https://github.com/kubernetes-sigs/kube-agentic-networking/pull/17#issuecomment-3601327274) design proposed by @david-martin would help attenuate concerns about auth mechanisms being too specific of one or another implementation?

I believe in the end what we all want is not something that makes sense to me, to you, or to any other who impersonates an implementation, but something that makes sense to _users_. I have a strong feeling that users will value anything that feels less implementation-specific and more in line with what the common denominator underneath (aka Kubernetes in this case) has to offer, of course as long as it solves their use cases. A prototype that is not tied to any particular implementation may be a good way to collect feedback on this idea IMO.

keithmattix replied:
Sorry, I have to agree with @howardjohn on this one. An API that makes sense to users is definitely our shared goal, but Kubernetes as an AuthZ API for network traffic falls short of that IMO. I don't think I've talked to a single user at Azure who has even thought about doing something like that. That's not to say this API shouldn't support such a thing as a vendor extension, but having an explicit API seems like a bit much

david-martin commented:
> I love the diagrams! thanks @david-martin

Thanks

> What would `providerRef` ref to? A provider specific GVK?

Yes.

> > I'd probably consider having the "inline rules" also as an extension type referenced via providerRef.
> > Why?
> 
> Any concerns with "simple" rules that are identified (or could be) as broadly applicable?

This was my thinking. All implementations ship with some simple solution that works for a broad set of use cases, but is perhaps not scalable and has some other limitations.

guicassolato commented:
PR updated based on recent feedback and discussions held at the community call and on Slack:

- Removed suggestion of having Kubernetes TokenReview and SubjectAccessReview authN/Z methods into the standard API

- Added support for external authz (reusing types defined by Gateway API)

- Dropped the `AuthScheme` CRD altogether to favour extending the `AccessPolicy` one instead, so inline static access control rules can be mixed and combined with offloading and pattern-matching ones into the same policy resource.

cc @LiorLieberman

haiyanmeng commented on docs/proposals/0017-DynamicAuth.md:
```diff
@@ -0,0 +1,441 @@
+
+## Non-Goals
+
+Agent and user authentication is not within the scope of this proposal.
```
This seems incorrect since the proposal does introduce a new field `OIDC` for authN.

haiyanmeng commented on docs/proposals/0017-DynamicAuth.md:
```diff
@@ -0,0 +1,441 @@
+	// The field is defined as the HTTPExternalAuthFilter type from
+	// Gateway API: https://pkg.go.dev/sigs.k8s.io/gateway-api/apis/v1#HTTPExternalAuthFilter
+	// +optional
+	ExternalAuth *gatewayapiv1.HTTPExternalAuthFilter `json:"externalAuth,omitempty"`
```
If `externalAuth` is used, is the `source` field still required?

hzxuzhonghu replied:
I feel the same `externalAuth` seems separate from the inner auth.

guicassolato commented on docs/proposals/0017-DynamicAuth.md:
```diff
@@ -0,0 +1,441 @@
+	// The request is expected to carry a valid access token issued the trusted authentication
+	// server in the Authorization: header
+	// +optional
+	OIDC *OIDC `json:"oidc,omitempty"`
```
This is for authenticating the requests that are sent by MCP clients to the MCP backend, period. The where depends on the implementation. It may be in a gateway, but it could also be in a sidecar proxy.

The part that maters here is the intent. By specifying OIDC in an AccessPolicy, the policy owner is stating that traffic directed to the targeted Backend, carrying in the `Authorization:` header of the request a valid OIDC token, issued by the trusted IdP specified in the `issuerUrl` field, must be deemed authorised.

piyushnigam replied:
My 2¢.

I think this proposal is trying to represent an principal selector mechanism to apply the authorization policy on. I think adding the OIDC internals here is a scope creep. I would propose a slightly different design that shoudl achieve something similar.

<Authentication Policy> => <Principal Selector Policy> => <Access Policy>

The principal selector policy should have a predefined order of precedence for producing an unambiguous principal. For example, mTLS -> Authorization header -> SAML -> plaintext hedaer. If there are multiple identities present, you can produce named principals, like defaultPrincipal=<mTLSidentity>, OIDCPrincipal=<OIDCIdentity>. Then the access policy can refer to one or more principals in the request to apply policies.

If we follow the model above, OIDC configuration is contained within the authentication policy. The principal selector can override the ordering / naming of the produced principals to apply to the access policy. But the access policy itself remain unaware of how authentication is being performed. It can just rely on the unambiguous principal being produced to apply policies. If there is a need for a different principal selection, the access policy can explicitly refer to it.

The current approach pollutes the AccessPolicy with OIDC parameters that do not fit well here as mentioned in the non-goals section.

/cc @LiorLieberman

guicassolato replied:
> The principal selector policy should have a predefined order of precedence for producing an unambiguous principal.

I'd be a +1 on having a predefined order of precedence between the different ways of expressing the source of the request – i.e., between SPIFE IDs, SAs, and OIDC servers, if more than one is specified in the same rule. Maybe even requiring those to be a one-of kind of field I'd be OK with, if we think this could help remove/mitigate any ambiguity.

> Then the access policy can refer to one or more principals in the request to apply policies.

I think we should stick with the one-principal-per-request approach, relying on the order of the rules and the precedence between methods, rather than having to deal with different kinds and/or named principals.


> […] adding the OIDC internals here is a scope creep […]
> […] the access policy itself remain unaware of how authentication is being performed. […]
> The current approach pollutes the AccessPolicy with OIDC parameters […]

The purpose of adding OIDC – or any other protocol-specific, authentication-related parameters in a policy – is for the user (who owns the policy and often, though not necessarily, who owns the targets of the policy too) to inform a PEP about what authentication method is already in place in the system, so the PEP can enforce zero-trust request verification accordingly.

Can you please explain how, by not informing the PEP that the authentication method being used by the requestors is an OIDC server X, the PEP can possibly know it has to lookup for OIDC ID tokens in the request to verify?

guicassolato commented on docs/proposals/0017-DynamicAuth.md:
```diff
@@ -0,0 +1,441 @@
+	Audiences []string `json:"audiences,omitempty"`
+	// Scopes is a list of acceptable scopes for the OIDC tokens
+	// +optional
+	Scopes []string `json:"scopes,omitempty"`
```
No. The OIDC token provided in the request must be issued for such scopes, which is usually identifiable by the checking the `scope` claim of the JWT.

BTW, if this was bare OAuth, not OIDC, checking the scope could in some cases require introspecting the token back sending an online request to the IdP. We want to avoid that by preferring a protocol that is guaranteed to include tokens that can be verified offline.

guicassolato commented on docs/proposals/0017-DynamicAuth.md:
```diff
@@ -0,0 +1,441 @@
+
+#### Defense in Depth
+
+- Combine AccessPolicy with AccessPolicy: Use AccessPolicy for identity verification and leverage AccessPolicy for additional resource-level authorization when appropriate
```
Nice catch! I will fix that. Thanks.

guicassolato commented on docs/proposals/0017-DynamicAuth.md:
```diff
@@ -0,0 +1,441 @@
+	// - request.body["tool-name"] in identity.authorized_tools
+	// - identity.group == "admin"
+	// +optional
+	CEL *CELAuthorization `json:"cel,omitempty"`
```
AFAIK, there's nothing magic in CEL to resolve the identity. Maybe, it can be used extract identity information from the request but that'd still to be validated.

The way I see it, CEL is for authorisation. Something needs to assign a value to the `identity` root binding as a first step. This could be statically assigned by the whatever resolves the identity and therefore is capable of matching it to one of the statically defined list entries (namely `identities` and `serviceAccounts`), or resolved to the payload of the OIDC token (i.e. the claims of the JWT) when this is the identity verification method of choice. Then assertions can be made on the value assigned to this variable using CEL.

LiorLieberman replied:
>In this particular example though, I personally find using source.identities in combination to CEL easier to read.
Plus one on this, it also changes the whole model, which is sources (identities - are presented under `source`). Why providing many ways to do one thing from start (or at all)? its confusing for the users.

haiyanmeng commented on docs/proposals/0017-DynamicAuth.md:
```diff
@@ -0,0 +1,441 @@
+Date: 27th November 2025
+Authors: guicassolato
+Status: Provisional
```
LOL. I was setting a bad example in proposal-0008. We can keep this as is. Or maybe add empty lines to make it look nicer. Totally up to you. I can send out a small PR to fix proposal-0008 later.

haiyanmeng commented on docs/proposals/0017-DynamicAuth.md:
````diff
@@ -0,0 +1,441 @@
+- extracting identity information from the request based on standard authentication protocols (e.g. OIDC), and
+- verification methods to authorize an agent-to-Backend request based on pattern-matching rules (e.g. using CEL).
+
+```go
````
Sounds good. Let us keep this as is. Thanks!

LiorLieberman commented on docs/proposals/0017-DynamicAuth.md:
```diff
@@ -0,0 +1,441 @@
+	Authorization []AuthorizationRule `json:"authorization,omitempty"`
+}
+
+// Source specifies the source of a request.
```
This seems out of date (I do know you have taken it from 0008), opened https://github.com/kubernetes-sigs/kube-agentic-networking/pull/34 to fix. 

will be great if you can align it from there, include some minor changes to serviceaccount sources and identity comments.

guicassolato replied:
Updated.

hzxuzhonghu commented on docs/proposals/0017-DynamicAuth.md:
```diff
@@ -0,0 +1,441 @@
+	Source Source `json:"source"`
+	// Authorization specifies a list of rules that at least one must match for access to be granted.
+	// +optional
+	Authorization []AuthorizationRule `json:"authorization,omitempty"`
```
I would suggest split out this struct, make Source -> Condition(which means AccessCondition ) ->  Target in horizon level
